source/scale/redis.rst

```diff
@@ -0,0 +1,109 @@
+
+  a. Enter **Redis Address** for the Redis server you set up earlier (e.g., ``redis.example.com:6379``).
+  b. (Optional) Enter **Redis Password** if your Redis server requires authentication.
+  c. (Optional) Enter **Redis Database** to specify which Redis database to use (-1 by default, which uses Redis's default database).
+
+4. Save your configuration and restart the Mattermost server.
+
+Alternatively, you can configure Redis in the ``config.json`` file:
+
+.. code-block:: json
+
+    "CacheSettings": {
+        "CacheType": "redis",
+        "RedisAddress": "redis.example.com:6379",
+        "RedisPassword": "",
+        "RedisCachePrefix": "",
+        "RedisDB": -1,
+        "DisableClientCache": false
+    }
+
+Deploy with AWS ElastiCache
+---------------------------
+
+For enterprise-scale deployments, AWS ElastiCache for Redis provides a fully managed Redis service that simplifies deployment, operation, and scaling. 
+
+When deploying Mattermost with AWS ElastiCache:
+
+1. Create an ElastiCache for Redis instance, choosing Redis OSS version 7.1 or later.
+2. Based on load testing results, we recommend starting with a ``cache.m7g.2xlarge`` instance for deployments supporting over 100,000 users.
+3. Configure your Mattermost server to connect to the ElastiCache endpoint.
+4. Monitor CPU utilization closely, as Redis is single-threaded and performance bottlenecks are typically CPU-bound rather than memory-bound.
+
+Performance considerations
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+When using Redis with Mattermost, consider the following performance insights:
+
+- Current load tests indicate that memory usage of Redis is relatively low, with CPU being the primary bottleneck because Redis is single-threaded.
+- When planning your Redis deployment, use CPU utilization as the primary metric for scaling decisions.
+- For extremely large deployments, consider sharding your Redis deployment to distribute the workload.
+
+Implementation details
+----------------------
+
+Mattermost uses the `rueidis <https://github.com/redis/rueidis>`_ library instead of the more popular `go-redis <https://github.com/redis/go-redis>`_, primarily because rueidis offers more performance optimizations out of the box, including:
+
+- Automatic pipelining of concurrent queries
+- Client-side caching
+- Additional performance tuning options like MaxFlushDelay
+
+The primary implementation files include:
+
+- ``platform/services/cache/provider.go`` - Contains the library initialization code
+- ``platform/services/cache/redis.go`` - Contains the cache interface implementation
+
+Frequently asked questions (FAQ)
+--------------------------------
+
+Do I need to use Redis?
+~~~~~~~~~~~~~~~~~~~~~~~
+
+While Redis is not required for smaller Mattermost deployments, it is highly recommended for large-scale installations, especially those with more than 100,000 users. Redis significantly improves performance by reducing database load and optimizing common operations.
```

Comment:
@agnivade - Are we saying there's no benefit to Redis sub 100k?

Reply:
Would you prefer to remove that line? Honestly speaking, there isn't really that much of an improvement at low scale.

Reply:
Maybe my interpretation is biased, but I don't see `not required` as meaning it's not beneficial at all. I think that having an extra component in a deployment is always a tradeoff to consider so we probably wouldn't recommend it for smaller deployments.

Reply:
@agnivade - Is the next step here to clarify that Redis isn't recommended for smaller deployments under 100K?

Reply:
@agnivade - I recommend being more explicit and prescriptive to ensure that smaller customers understand whether Redis applies to their scenario.

Reply:
Understood. How about something like this:

```
- If your installation has more than 100K users, Redis is highly recommended.
- If your installation has less than 100K users, Redis is not recommended.
```
?

Reply:
@agnivade I know we typically refer to "users" for simplicity, but I find it unrealistic for customers to think in such strict terms. For example, what if a customer has 80,000 users? Perhaps we could suggest that above 50k users (just a random example), it's recommended to start considering Redis, and above 100k users, it becomes highly recommended.

More broadly, do you think we could define a few self-diagnostic guidelines for when Redis might become necessary, for example, by identifying specific thresholds in Grafana?

Reply:
I believe Carrie was looking to have a more prescriptive guidance for the customers. If we introduce a third level into this (not needed, start thinking, absolutely needed), it might get complicated.

Re: specific thresholds - that's a great point. Yes, it's a correlation we are drawing with number of users with scale. Specifically, in this case, the number of nodes in your cluster, and rate of cache invalidations would be the threshold to monitor. I can come up with the exact prom query for that and update the PR.

Reply:
> If we introduce a third level into this (not needed, start thinking, absolutely needed), it might get complicated.

Yep, I completely understand. I just feel that making it a strict yes/no decision based on an arbitrary number (like 100k) doesn't add much value either. I would prefer to provide guidelines to help customers identify when it might be time to consider using Redis.

> Specifically, in this case, the number of nodes in your cluster, and rate of cache invalidations would be the threshold to monitor. I can come up with the exact prom query for that and update the PR.

That would be amazing :pray:

Reply:
PTAL @streamer45